fw/http_parser.c

```diff
@@ -5957,36 +5962,45 @@ do {									\
  * besides, @str must be of plain @TfwStr{} type and variable @fld is
  * used (instead of hard coded header field).
  */
-#define H2_TRY_STR_2LAMBDA_fixup(str, fld, lambda1, lambda2, curr_st, next_st) \
-	BUG_ON(!TFW_STR_PLAIN(str));					\
-	if (!chunk->data)						\
-		chunk->data = p;					\
-	__fsm_n = __try_str(fld, chunk, p, __data_remain(p),		\
-			    (str)->data, (str)->len);			\
-	if (__fsm_n > 0) {						\
-		if (likely(chunk->len == (str)->len)) {			\
-			lambda1;					\
-			TRY_STR_INIT();					\
-			__msg_field_fixup_pos(fld, p, __fsm_n);		\
-			__FSM_I_field_chunk_flags(fld, TFW_STR_HDR_VALUE); \
-			if (__data_off(p + __fsm_n) < len) {		\
-				p += __fsm_n;				\
-				goto next_st;				\
-			}						\
-			if (likely(fin))				\
-				lambda2;				\
-			parser->_i_st = &&next_st;			\
-			__FSM_EXIT(CSTR_POSTPONE);			\
-		}							\
-		if (likely(fin))					\
-			__FSM_EXIT(CSTR_NEQ);				\
-		parser->_i_st = &&curr_st;				\
-		__msg_field_fixup_pos(fld, p, __fsm_n);			\
-		__FSM_I_field_chunk_flags(fld, TFW_STR_HDR_VALUE);	\
-		__FSM_EXIT(CSTR_POSTPONE);				\
-	}
-
-#define H2_TRY_STR_LAMBDA_fixup(str, fld, lambda, curr_st, next_st)	\
+#define H2_TRY_STR_3LAMBDA_fixup(str, fld, lambda1, lambda2, lambda3,           \
+				 curr_st, next_st)				\
+	BUG_ON(!TFW_STR_PLAIN(str));						\
+	if (!chunk->data)							\
+		chunk->data = p;						\
+	__fsm_n = __try_str(fld, chunk, p, __data_remain(p),			\
+			    (str)->data, (str)->len);				\
+	do {									\
+		if (__fsm_n > 0) {						\
+			if (likely(chunk->len == (str)->len)) {			\
+				lambda1;					\
+				TRY_STR_INIT();					\
+				__msg_field_fixup_pos(fld, p, __fsm_n);		\
+				__FSM_I_field_chunk_flags(fld, TFW_STR_HDR_VALUE); \
+				if (__data_off(p + __fsm_n) < len) {		\
+					p += __fsm_n;				\
+					goto next_st;				\
+				}						\
+				if (likely(fin))				\
+					lambda2;				\
+				parser->_i_st = &&next_st;			\
+				__FSM_EXIT(CSTR_POSTPONE);			\
+			}							\
+			__msg_field_fixup_pos(fld, p, __fsm_n);			\
+			__FSM_I_field_chunk_flags(fld, TFW_STR_HDR_VALUE);	\
```

Comment:
Yes, you're right: we have such lambda1 in `__h2_req_parse_content_type()` state `I_ContTypeParam`. Could you please add a comment for `lambda1` in `H2_TRY_STR_3LAMBDA_fixup` that it might return with an error code and we don't need to fixup the current data chunk.

Reply:
Done